### PR TITLE
Add custom value in SelectBox on enter key pressing (T710656)

### PR DESCRIPTION
**Pull request by KsyushaM**

### CODE REVIEW

File: js/ui/select_box.js [hunk 1]
```diff
@@ -109,7 +109,11 @@ var SelectBox = DropDownList.inherit({
                 this._cancelEditing();
             },
             enter: function(e) {
-                if(this._input().val() === "" && this.option("value") && this.option("allowClearing")) {
+                var inputText = this._input().val();
+                var inputTextIsEmpty = inputText === "";
```
Comment (DokaRus) [inline js/ui/select_box.js]: [`stringUtils.isEmpty`](https://github.com/DevExpress/DevExtreme/blob/18_2/js/core/utils/string.js#L91)?

File: js/ui/select_box.js [hunk 2]
```diff
@@ -109,7 +109,11 @@ var SelectBox = DropDownList.inherit({
...
+                var customInput = this._list && !this._list.option("focusedElement");
```
Comment (DokaRus) [inline js/ui/select_box.js]: It isn't an input. It is a list element.

File: js/ui/select_box.js [hunk 3]
```diff
@@ -119,6 +123,12 @@ var SelectBox = DropDownList.inherit({
                 } else {
                     if(this.option("acceptCustomValue")) {
                         e.preventDefault();
+
+                        if(!inputTextIsEmpty && customInput) {
```
Comment (DokaRus) [inline js/ui/select_box.js]: Should we compare a focused list item text and input value?

File: js/ui/select_box.js (hunk 2 above)
Comment (DokaRus) [reply js/ui/select_box.js]: Should it be something like `hasCustomValue`?

File: js/ui/select_box.js [hunk 4]
```diff
@@ -109,7 +110,11 @@ var SelectBox = DropDownList.inherit({
                 this._cancelEditing();
             },
             enter: function(e) {
-                if(this._input().val() === "" && this.option("value") && this.option("allowClearing")) {
+                var inputText = this._input().val();
+                var inputTextIsEmpty = isEmpty(inputText);
+                var inputHasCustomValue = this._list && !this._list.option("focusedElement");
```
Comment (mrjohn0011) [inline js/ui/select_box.js]: What if `focusStateEnabled` is false? All mobile devices have no focusedElement in the list. Please check this

**Review by mrjohn0011 — CHANGES_REQUESTED** [review]
Write one more test when the editor is initially closed. Should the item be added after enter pressed?

File: js/ui/select_box.js [hunk 5]
```diff
@@ -119,6 +124,12 @@ var SelectBox = DropDownList.inherit({
                 } else {
                     if(this.option("acceptCustomValue")) {
                         e.preventDefault();
+
```
Comment (churkin) [inline js/ui/select_box.js:126]:
```javascript
var isOpened = this.option("opened");
```

File: js/ui/select_box.js [hunk 6]
```diff
@@ -109,7 +110,11 @@ var SelectBox = DropDownList.inherit({
                 this._cancelEditing();
             },
             enter: function(e) {
-                if(this._input().val() === "" && this.option("value") && this.option("allowClearing")) {
+                var inputText = this._input().val();
```
Comment (churkin) [inline js/ui/select_box.js]: `this._input().val().trim()`

File: js/ui/select_box.js [hunk 7]
```diff
@@ -109,7 +110,11 @@ var SelectBox = DropDownList.inherit({
...
+                var inputTextIsEmpty = isEmpty(inputText);
```
Comment (churkin) [inline js/ui/select_box.js]: remove

File: js/ui/select_box.js [hunk 8]
```diff
@@ -109,7 +110,11 @@ var SelectBox = DropDownList.inherit({
...
+                var inputHasCustomValue = this._list && !this._list.option("focusedElement");
+
+                if(inputTextIsEmpty && this.option("value") && this.option("allowClearing")) {
```
Comment (churkin) [inline js/ui/select_box.js]:
```javascript
if(!inputText && this.option("value") && this.option("allowClearing")) {
```

File: js/ui/select_box.js [hunk 9]
```diff
@@ -119,6 +124,12 @@ var SelectBox = DropDownList.inherit({
                 } else {
                     if(this.option("acceptCustomValue")) {
                         e.preventDefault();
+
+                        if(!inputTextIsEmpty && inputHasCustomValue) {
```
Comment (churkin) [inline js/ui/select_box.js]:
```javascript
 if(isCustomText) {
```

File: js/ui/select_box.js (hunk 4 above)
Comment (churkin) [inline js/ui/select_box.js]:
```javascript
var isCustomText = inputText && this._list && !this._list.option("focusedElement");
```

**Review by churkin — CHANGES_REQUESTED** [review]

File: js/ui/select_box.js (hunk 6 above)
Comment (KsyushaM) [reply js/ui/select_box.js]: If customer types " ", it will be added to list. Is it OK?

**Review by mrjohn0011 — APPROVED** [review]

**Review by churkin — APPROVED** [review]

**Pull request merged**